encode/vp8enc.c

```diff
@@ -0,0 +1,1074 @@
+          break;
+      case 7:
+          tmp_input = atoi(optarg);
+          parameter_check("--hrd_win",tmp_input,1000,8000);
+          settings.hrd_window = tmp_input;
+          break;
+      case 8:
+          tmp_input = atoi(optarg);
+          parameter_check_positive("--vbr_max",tmp_input,1);
+          settings.max_variable_bitrate = tmp_input;
+          break;
+      case 9:
+          tmp_input = atoi(optarg);
+          parameter_check_positive("--fn_num" ,tmp_input,1);
+          settings.num_frames = tmp_input;
+          break;
+      case 10:
+          settings.error_resilient = 1;
+          break;
+      case 11:
+          settings.debug = 1;
+          break;
+      case 'h':
+      case 0:
+      default:
+          return PARSE_OPTIONS_FAIL;
+          break;
+    }
+  }
+  return PARSE_OPTIONS_OK;
+}
+
+int main(int argc, char *argv[])
+{
+  int current_frame, frame_type;
+  FILE *fp_vp8_output, *fp_yuv_input;
+  uint64_t timestamp;
+  struct timeval t1,t2;
+  double fps, elapsed_time;
+
+
+  if(argc < 5) {
+      vp8enc_show_help();
+      return VP8ENC_FAIL;
+  }
+
+  if(parse_options(argc-4, &argv[4]) != PARSE_OPTIONS_OK)
+  {
+    vp8enc_show_help();
+    return VP8ENC_FAIL;
+  }
+
+  settings.width = atoi(argv[1]);
+  parameter_check("Width", settings.width, 16, MAX_XY_RESOLUTION);
+
+  settings.height = atoi(argv[2]);
+  parameter_check("Height", settings.height, 16, MAX_XY_RESOLUTION);
+
+  fp_yuv_input = fopen(argv[3],"rb");
+  if(fp_yuv_input == NULL)
+  {
+    fprintf(stderr,"Error: Couldn't open input file.\n");
+    return VP8ENC_FAIL;
+  }
+
+  fp_vp8_output = fopen(argv[4],"wb");
+  if(fp_vp8_output == NULL)
+  {
+    fprintf(stderr,"Error: Couldn't open output file.\n");
+    return VP8ENC_FAIL;
+  }
+
+  if( settings.rc_mode == VA_RC_VBR && settings.max_variable_bitrate < settings.frame_bitrate)
+  {
+      fprintf(stderr,"Error: max. variable bitrate should be greater than frame bitrate (--vbr_max >= --fb)\n");
+      assert(0);
+  }
+
+  settings.frame_size = settings.width * settings.height * 3 / 2; //NV12 Colorspace - For a 2x2 group of pixels, you have 4 Y samples and 1 U and 1 V sample.
+  if(!settings.num_frames)
+    settings.num_frames = vp8enc_get_FileSize(fp_yuv_input)/(size_t)settings.frame_size;
+  settings.codedbuf_size = settings.width * settings.height; //just a generous assumptions
+
+  fprintf(stderr,"Info: Encoding total of %d frames.\n",settings.num_frames);
+
+  gettimeofday(&t1,0); //Measure Runtime
+
+  vp8enc_init_VaapiContext();
+  vp8enc_create_EncoderPipe();
+
+  vp8enc_write_ivf_header(fp_vp8_output);
+
+  current_frame = 0;
+  timestamp = 0;
+  while (current_frame < settings.num_frames)
+  {
+    fprintf(stderr,"\rProcessing frame: %d",current_frame);
+
+    if ( (current_frame%settings.intra_period) == 0)
+      frame_type = KEY_FRAME;
+    else
+      frame_type = INTER_FRAME;
+
+    vp8enc_upload_yuv_to_surface(fp_yuv_input, vaapi_context.recon_surface, current_frame);
```

Comment:
It seems you are trying to upload the raw data to the reconstructed surface which is wrong.
You are supposed to provide a free surface id as reconstructed surface and the input surface should be only submitted through vaBeginPicture.

Reply:
I will introduce a surface dedicated for the input.